File: 0940-fruit-into-baskets/0940-fruit-into-baskets.java
class Solution {
    public int totalFruit(int[] fruits) {
        HashMap<Integer,Integer>H=new HashMap<>();
        int l=0;
        int n=fruits.length;
        int ans=0;
        for(int r=0;r<n;r++){
            int val=fruits[r];
            H.put(val,H.getOrDefault(val,0)+1);
           while (H.size() > 2) {
                int lval = fruits[l];
                if (H.containsKey(lval)) {  
                    H.put(lval, H.get(lval) - 1); 
                    if (H.get(lval) == 0) {
                        H.remove(lval);  
                    }
                }
                l++;
                 
            }
            ans=Math.max(ans,r-l+1);
        }
        return ans;
    }
}

File: 0940-fruit-into-baskets/0940-fruit-into-baskets.py
class Solution:
    def totalFruit(self, fruits: List[int]) -> int:
        counter = Counter()
        l, r = 0, 0
        ans = 0
        for r in range(len(fruits)):
            counter[fruits[r]] += 1
            while len(counter) == 3:
                counter[fruits[l]] -= 1
                if counter[fruits[l]] == 0:
                    # remove key to decrease len(counter)
                    del counter[fruits[l]] 
                l += 1
            ans = max(ans, r + 1 - l)
        return ans
            
